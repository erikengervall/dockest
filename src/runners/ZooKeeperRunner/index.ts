import {
  BaseRunner,
  GetComposeService,
  Runner,
  SharedDefaultableConfigProps,
  SharedRequiredConfigProps,
} from '../@types'
import { DEFAULT_CONFIG_PROPS, SHARED_DEFAULT_CONFIG_PROPS } from '../constants'
import { ObjStrStr } from '../../@types'
import ConfigurationError from '../../errors/ConfigurationError'
import getDependsOn from '../utils/getDependsOn'
import getImage from '../utils/getImage'
import getPorts from '../utils/getPorts'
import Logger from '../../Logger'
import validateConfig from '../../utils/validateConfig'
import validateTypes from '../../utils/validateTypes'

type RequiredConfigProps = {
  service: string
} & SharedRequiredConfigProps

type DefaultableConfigProps = {
  commands: string[]
  connectionTimeout: number
  dependsOn: Runner[]
  host: string
  image: string | undefined
  ports: ObjStrStr
} & SharedDefaultableConfigProps

type ZooKeeperRunnerConfig = RequiredConfigProps & DefaultableConfigProps

const DEFAULT_PORT = '2181'
const DEFAULT_CONFIG: DefaultableConfigProps = {
  ...SHARED_DEFAULT_CONFIG_PROPS,
  commands: DEFAULT_CONFIG_PROPS.COMMANDS,
  connectionTimeout: DEFAULT_CONFIG_PROPS.CONNECTION_TIMEOUT,
  dependsOn: DEFAULT_CONFIG_PROPS.DEPENDS_ON,
  host: DEFAULT_CONFIG_PROPS.HOST,
  image: DEFAULT_CONFIG_PROPS.IMAGE,
  ports: {
    [DEFAULT_PORT]: DEFAULT_PORT,
  },
}

class ZooKeeperRunner implements BaseRunner {
  public static DEFAULT_HOST = DEFAULT_CONFIG_PROPS.HOST
  public static DEFAULT_PORT = DEFAULT_PORT
  public containerId = ''
  public initializer = ''
  public runnerConfig: ZooKeeperRunnerConfig
  public logger: Logger

  public constructor(config: RequiredConfigProps & Partial<DefaultableConfigProps>) {
    this.runnerConfig = {
      ...DEFAULT_CONFIG,
      ...config,
    }
    this.logger = new Logger(this)

    const schema: { [key in keyof RequiredConfigProps]: any } = {
      service: validateTypes.isString,
    }
    validateConfig(schema, this.runnerConfig)
  }

  public getComposeService: GetComposeService = composeFileName => {
    const { dependsOn, image, ports, service } = this.runnerConfig

    const ZOOKEEPER_CLIENT_PORT = Object.keys(ports).find(key => ports[key] === DEFAULT_PORT)
    if (!ZOOKEEPER_CLIENT_PORT) {
      throw new ConfigurationError(
        `Could not resolve required environment variable ZOOKEEPER_CLIENT_PORT. Expected ${DEFAULT_PORT} to appear as value in ports object`,
      )
    }

    return {
      [service]: {
        environment: {
          ZOOKEEPER_CLIENT_PORT,
        },
        ...getDependsOn(dependsOn),
        ...getImage({ image, composeFileName, service }),
        ...getPorts(ports),
      },
    }
  }
}

export { ZooKeeperRunnerConfig }
export default ZooKeeperRunner
