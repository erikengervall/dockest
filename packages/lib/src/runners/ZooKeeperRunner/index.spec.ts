/* eslint-disable @typescript-eslint/ban-ts-ignore */

import ZooKeeperRunner from './index'

const ZooKeeperRunner1 = new ZooKeeperRunner({ service: 'zk1', image: 'some/image:123' })
const ZooKeeperRunner2 = new ZooKeeperRunner({ service: 'zk2', image: 'some/image:123' })

describe('ZooKeeperRunner', () => {
  it('should create unique instances', () => {
    expect(ZooKeeperRunner1).not.toBe(ZooKeeperRunner2)
    expect(ZooKeeperRunner1).toMatchSnapshot()
    expect(ZooKeeperRunner2).toMatchSnapshot()
  })

  it('should fail validation', () => {
    // @ts-ignore
    expect(() => new ZooKeeperRunner({}).validateConfig()).toThrow(/service: Schema-key missing in config/)
  })
})
