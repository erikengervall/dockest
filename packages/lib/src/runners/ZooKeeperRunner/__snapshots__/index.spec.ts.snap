// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZooKeeperRunner should create unique instances 1`] = `
ZooKeeperRunner {
  "containerId": "",
  "getComposeService": [Function],
  "initializer": "",
  "isBridgeNetworkMode": false,
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "zk1",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "ports": Array [
      Object {
        "published": 2181,
        "target": 2181,
      },
    ],
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "zk1",
  },
}
`;

exports[`ZooKeeperRunner should create unique instances 2`] = `
ZooKeeperRunner {
  "containerId": "",
  "getComposeService": [Function],
  "initializer": "",
  "isBridgeNetworkMode": false,
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "zk2",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "ports": Array [
      Object {
        "published": 2181,
        "target": 2181,
      },
    ],
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "zk2",
  },
}
`;
